server/jobs_controller.go

```diff
@@ -0,0 +1,201 @@
+	"github.com/runatlantis/atlantis/server/logging"
+)
+
+// Concurrency model in this controller:
+//
+// - We keep a buffer of all the output so far for each PR
+// - We keep a set of clients for each PR
+// - When a new client is added, the entire buffer is sent through so
+//   it is "up to date"
+// - When a new line of output is added, we send it to all current
+//   clients and then add it to the buffer
+// - Changes to the buffer or set of clients are wrapped in a mutex
+//
+// This maintains the invariant that whenever a function call returns,
+// all of the clients have seen every message that is in the buffer.
+//
+// When we get a message that tells us to clear the buffer, we just do
+// it, without telling any clients about it. This won't have any
+// effect on the frontend, except that newly connected clients won't
+// get the old messages sent to them, because the server has
+// "forgotten" about them. We clear the buffer for a PR when starting
+// and finishing processing that PR, because after the processing has
+// finished the output is posted to GitHub anyway and it doesn't need
+// to be persisted in the Atlantis web UI.
+
+type JobsController struct {
+	AtlantisVersion     string
+	AtlantisURL         *url.URL
+	Logger              *logging.SimpleLogger
+	JobTemplate         TemplateWriter
+	TerraformOutputChan <-chan *models.TerraformOutputLine
+
+	logBuffers map[string][]string
+	wsChans    map[string]map[chan string]bool
+	chanLock   sync.Mutex
+}
+
+// Register client to get Terraform output for PR. Send all the output
+// currently in the buffer.
+func (j *JobsController) addChan(pull string) chan string {
+	ch := make(chan string, 1000)
+	j.chanLock.Lock()
+	for _, line := range j.logBuffers[pull] {
+		ch <- line
+	}
+	if j.wsChans == nil {
+		j.wsChans = map[string]map[chan string]bool{}
+	}
+	if j.wsChans[pull] == nil {
+		j.wsChans[pull] = map[chan string]bool{}
+	}
+	j.wsChans[pull][ch] = true
+	j.chanLock.Unlock()
+	return ch
+}
+
+// Deregister client to no longer get Terraform output for PR.
+func (j *JobsController) removeChan(pull string, ch chan string) {
+	j.chanLock.Lock()
+	delete(j.wsChans[pull], ch)
+	j.chanLock.Unlock()
+}
+
+// Clear out log lines that are in the buffer (which would be
+// immediately received by a newly connecting client). This won't
+// cause anything to be deleted from an existing client webpage, but
+// it does mean the data won't reappear on a refresh.
+func (j *JobsController) clearLogLines(pull string) {
+	j.chanLock.Lock()
+	delete(j.logBuffers, pull)
+	j.chanLock.Unlock()
+}
+
+// Add a log line to the buffer and send it to all clients.
+func (j *JobsController) writeLogLine(pull string, line string) {
+	j.chanLock.Lock()
+	if j.logBuffers == nil {
+		j.logBuffers = map[string][]string{}
+	}
+	for ch, _ := range j.wsChans[pull] {
+		select {
+		case ch <- line:
+		default:
+			// If we get to this case then it means that
+			// the channel is full, which probably means
```

Comment:
This channel is unbuffered though? How would it be full.

Reply:
I was referring to this channel: https://github.com/runatlantis/atlantis/pull/1315/files#diff-624efb8be6c7fd9ea8c6001e775c83da7640f9fb7bba9b19f71e4ee92dd803b0R52